R/get_p_value.R

```diff
@@ -1,95 +1,115 @@
-#'   specify(mpg ~ am) %>% 
+#'   specify(mpg ~ am) %>%
 #'   hypothesize(null = "independence") %>%
 #'   generate(reps = 100) %>%
 #'   calculate(stat = "diff in means", order = c("1", "0"))
-#' null_distn %>% 
-#'   p_value(obs_stat = d_hat, direction = "right")
-#' 
-#' @name get_pvalue
+#'
+#' # What proportion of replicates had a difference
+#' # in means more extreme than in the dataset?
+#' null_distn %>%
+#'   get_p_value(obs_stat = d_hat, direction = "right")
+#' @name get_p_value
 NULL
 
-#' @rdname get_pvalue
+#' @rdname get_p_value
 #' @export
-p_value <- function(x, obs_stat, direction){
-  
+get_p_value <- function(x, obs_stat, direction){
+
   check_type(x, is.data.frame)
+  if(!is_generated(x) & is_hypothesized(x)) {
+    stop_glue(
+      "Theoretical p-values are not yet supported.",
+      "`x` should be the result of calling `generate()`.",
+      .sep = " "
+    )
+  }
   obs_stat <- check_obs_stat(obs_stat)
   check_direction(direction)
-  
-  is_simulation_based <- !is.null(attr(x, "generate")) &&
-    attr(x, "generate")
-
-  if(is_simulation_based)
-    pvalue <- simulation_based_p_value(x = x, obs_stat = obs_stat, 
-                                        direction = direction)
-  
+
+  pvalue <- simulation_based_p_value(
+    x = x, 
+    obs_stat = obs_stat,
+    direction = direction)
+
   ## Theoretical-based p-value
   # Could be more specific
   # else if(is.null(attr(x, "theory_type")) || is.null(attr(x, "distr_param")))
   #   stop_glue("Attributes have not been set appropriately. ",
   #             "Check your {{infer}} pipeline again.")
-  
+
   # if(!("stat" %in% names(x))){
   #    # Theoretical distribution
-  #  which_distribution(x, 
+  #  which_distribution(x,
   #                     theory_type <- attr(x, "theory_type"),
   #                     obs_stat = obs_stat,
-  #                     direction = direction) 
+  #                     direction = direction)
   # }
-  
+
   return(pvalue)
 }
 
+#' @rdname get_p_value
+#' @export
+get_pvalue <- function(x, obs_stat, direction) {
+  get_p_value(x = x, obs_stat = obs_stat, direction = direction)
+}
+
 simulation_based_p_value <- function(x, obs_stat, direction){
-  
+
   if(direction %in% c("less", "left")){
-   p_value <- x %>% 
+    p_value <- x %>%
       dplyr::summarize(p_value = mean(stat <= obs_stat))
   }
   else if(direction %in% c("greater", "right")){
-   p_value <- x %>% 
+    p_value <- x %>%
       dplyr::summarize(p_value = mean(stat >= obs_stat))
   }
   else{
     p_value <- x %>% two_sided_p_value(obs_stat = obs_stat)
   }
-  
+
   p_value
 }
 
 two_sided_p_value <- function(x, obs_stat){
-  
+
   if(stats::median(x$stat) >= obs_stat){
     basic_p_value <- get_percentile(x$stat, obs_stat) +
-      (1 - get_percentile(x$stat, stats::median(x$stat) + 
-                       stats::median(x$stat) - obs_stat))
+      (1 - get_percentile(x$stat, stats::median(x$stat) +
+          stats::median(x$stat) - obs_stat))
```

Comment:
There are several ways to calculate a simulation/permutation based two-sided p-value. Do you know where this one came from? Here is what is used in Chihara and Hesterberg: "For a two-sided test, we calculate both one-sided P-values, multiply the smaller by 2, and finally (if necessary) round down to 1.0 (because probabilities can never be larger than 1.0)."

Reply:
I'm much more in favor of the Chihara and Hesterberg approach here as mine was a bit of a hack that doesn't actually work well if the distribution deviates from symmetry. We'll have to make sure it also aligns with how we shade with `visualize()` and the `shade_*` functions too whatever we decide. Can you take a stab at implementing Chihara and Hesterberg's? I'm happy to review but my time is limited over the next few weeks.